README.md
# Item-Catalog

The **Item-Catalog** is a project written for **Udacity's Full Stack Developer Nanodegree Program**. 

In this project, the webpage will display a menu and an item pane populated with food items with its representative food category. You have the option of adding new items to a particular food category, and edit/delete food items if you are the creator of that object (or have admin privileges). To create/edit/delete a category, you must have admin privileges. For the purpose of this evaluation, I have created a function to allow you to alternate your privilege status (See Usage.)

## Requirments

Please ensure your system contains the following software prior to using this project:

* Virtual Box
* Custom Vagrant file from Udacity.com

## Download

To obtain a copy of this project download the entire contents of this repository in a `.zip` file onto your desktop or folder of choice and unzip it to your vagrant folder. Below are also the required links that you will need to download and install for the requirements.

* [Vagrant](https://www.vagrantup.com/downloads.html)
* [Vagrant Configuration File](https://d17h27t6h515a5.cloudfront.net/topher/2017/August/59822701_fsnd-virtual-machine/fsnd-virtual-machine.zip)

## Usage

1) Open terminal or command prompt and navigate to your Vagrant folder.

2) Enter `vagrant up` to initially set up your virtual machine.

3) Once Vagrant has finished setting up, enter `vagrant ssh`. 

4) Navigate to your vagrant folder inside vagrant using `cd /vagrant/`.

5) This project has already been setup with a database. If you want to delete the database and recreate it, move `main.db` to the trash. Run `python database_setup.py` for a clean database or `python addDB.py` for a populated database.

6) Enter `python /vagrant/main.py` to run the server.

7) To view the webpage, visit: [localhost:5000].

[localhost:5000]: localhost:5000

8) To alternate between user and admin privileges, please sign into your Google Account first, then after successful completion visit: [localhost:5000/status/]

[localhost:5000/status/]: localhost:5000/status/

9) To serialize data visit:
- categories: `localhost:5000/main/JSON`
- items: `localhost:5000/main/<int:category_id>/item/JSON`
- specific item: `localhost:5000/main/<int:category_id>/item/<int:item_id>/JSON`
- users: `localhost:5000/user/JSON`

## Source

## Disclaimer

Please use this project at your own risk. I, _J. Ye._, am not responsible for any damage(s) that the end-user's computer may experience while using this project.

## License



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [v0.8.0-rc] - 2018-02-11
### Added
- Added `README.md` for instructions on how to use the project.
- Added statusChange() function to `main.py` to allow evaluators to alernate between user privileges, for testing purposes.

### Changed
- Modified `main.py`, `database_setup.py`, and `addDB.py` to be PEP8 compliant.

## [v0.7.0] - 2018-02-11
### Added
- Added createCategory(), editCategory(), deleteCategory(), newItem(), editItem(), deleteItem() functions to their appropriate locations in `main.html`.
- Added dropdown option in `newItem.html` to be able change what category you want to add the new item to.

### Changed
- Some webpages now will redirect back to `main.html` upon clicking **Item-Catalog**, others will redirect going back a page.
- Changed the text and url from login to logout when a user is logged in.

## Fixed
- Fixed Latest Entries display in `main.html`.

## [v0.6.0] - 2018-02-10
### Added
- Added `TODO.md` to keep track of progress and what features needs to be added before stable relase.

### Changed
- Modified `main.css` to clicking on Item-Catalog on the header pane will redirect back to `main.html`.
- Modified `addDB.py` to display more meaningful Category and Item entries for testing.

### Removed
- Removed `itemmenu.html`.

## [v0.5.0] - 2018-02-07
### Added
- Added `CHANGELOG.md` documenting all changes and revisions.
- Added version tags to commits. 
- Added login/logout functions in `login.html` and `main.py` using Google OAuth api.
- Added protection in `main.py` to pages from unauthorized data entry/edit/deletion.

### Changed
- Changed commit: a0033e8 to its correct version, [v0.2.2].
- Updated tables in `database_setup.py` to incorporate User table and it's functions.
- Updated functions in `main.py` to incorporate login features.

### Removed
- Removed commit(s): d1d2d2b, 15e8466, c9861a2 from Master. These were
erronous commits in attempts to adding the `.gitignore` file in [v0.2.2].
- Removed `logout.html` file.

## [v0.4.0] - 2018-02-07
### Added
- Added jsonify functions and serializable properties to `main.py` and `database_setup.py`, respectively.
- Added preliminary login state to `main.py`.
- Added functionality to newCategory, deleteCategory, itemMenu, editItem, and deleteItem.

### Changed
- Removed Test User from `addDB.py`.
- Updated `delete.html` to include deleteItem functionality.
- Renamed `new.html`, `itemmenu.html`, `edit.html`, `delete.html` to its corresponding redirect_uri. Also added additional `.html` files necessary for other redirect_uris.
- Updated all the templates to perform it's corresponding functions.
- Minor revisions to `main.css`.

## [v0.3.1] - 2018-02-06
### Added
- Added editItem and deleteItem functions to `main.py`.

### Changed
- Updated `delete.html` to include deleteItem functionality.

### Fixed
- Minor bug fixes and corrections to `main.html`.

## [v0.3.0] - 2018-02-06
### Added
- Added `addDB.py` to test the program using a test database.
- Added `main.css` for preliminarily styling and layout of the webpages.

### Changed
- Modified `main.html` to display category and item entries, and map certain app routes.
This webpage was also modified to include `main.css` as it's styling sheet.
- Modified `main.py` for related functions to render templates for `main.html`.

## [v0.2.2] - 2018-02-05
### Added
- Added `.gitignore` to not include DS_Store.

## [v0.2.1] - 2018-02-05
### Added
- Added template files: `new.html`, `login.html`, `logout.html`, `itemmenu.html`,
`edit.html`, `delete.html`. These files however have no functions and 
serve only as placeholders.

- Added preliminary layout of information to be displayed on `main.html`.

### Changed
- Minor code revisions to `main.py`.

## [v0.2.0] - 2018-01-26
### Fixed
- Major bug fixes to `database_setup.py` and `main.py` that prevented initial 
release from executing.

## v0.1.0 - 2018-01-26
### Added
- Added initial database setup to `database_setup.py` file.
- Added routing and corresponding functions to `main.py` file.
- Added empty template to `main.html` file.

[Unreleased]: https://github.com/jye0325/Item-Catalog/compare/v0.7.0...HEAD
[v0.7.0]: https://github.com/jye0325/Item-Catalog/compare/v0.6.0...v0.7.0
[v0.6.0]: https://github.com/jye0325/Item-Catalog/compare/v0.5.0...v0.6.0
[v0.5.0]: https://github.com/jye0325/Item-Catalog/compare/v0.4.0...v0.5.0
[v0.4.0]: https://github.com/jye0325/Item-Catalog/compare/v0.3.1...v0.4.0
[v0.3.1]: https://github.com/jye0325/Item-Catalog/compare/v0.3.0...v0.3.1
[v0.3.0]: https://github.com/jye0325/Item-Catalog/compare/v0.2.2...v0.3.0
[v0.2.2]: https://github.com/jye0325/Item-Catalog/compare/v0.2.1...v0.2.2
[v0.2.1]: https://github.com/jye0325/Item-Catalog/compare/v0.2.0...v0.2.1
[v0.2.0]: https://github.com/jye0325/Item-Catalog/compare/v0.1.0...v0.2.0

addDB.py
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker

from database_setup import User, Category, Item, Base

engine = create_engine('sqlite:////var/www/ItemCatalog/ItemCatalog/main.db')
# Bind the engine to the metadata of the Base class so that the
# declaratives can be accessed through a DBSession instance
Base.metadata.bind = engine

DBSession = sessionmaker(bind=engine)
# A DBSession() instance establishes all conversations with the database
# and represents a "staging zone" for all the objects loaded into the
# database session object. Any change made against the objects in the
# session won't be persisted into the database until you call
# session.commit(). If you're not happy about the changes, you can
# revert all of them back to the last commit by calling
# session.rollback()
session = DBSession()

# Category/Item Setup
category1 = Category(name="Chinese Cuisine")
session.add(category1)
session.commit()

categoryItem1 = Item(
    name="Soup Dumplings",
    description="Xiaolongbao is a type of Chinese steamed bun (baozi) from the\
     Jiangnan region, especially associated with Shanghai and Wuxi. It is \
    traditionally prepared in xiaolong, small bamboo steaming baskets, \
    which give them their name. Xiaolongbao are often referred to as a kind of\
     ""dumpling"", but should not be confused with British or American-style\
      dumplings, nor with Chinese jiaozi. Similarly, they are considered a\
       kind of ""soup dumpling"" but should not be confused with other \
       larger varieties of tang bao. In Shanghainese, they are also sometimes\
        known as siaulon moedeu or xiaolong-style mantous",
    category=category1
    )
session.add(categoryItem1)
session.commit()

categoryItem2 = Item(
    name="Lo Mein",
    description="Lo mein is a Chinese dish with wheat flour noodles. It often\
     contains vegetables and some type of meat or seafood, usually beef,\
      chicken, pork, shrimp or wontons. It can also be eaten with just\
       vegetables.",
    category=category1
    )
session.add(categoryItem2)
session.commit()

category2 = Category(name="Spanish Cuisine")
session.add(category2)
session.commit()

categoryItem1 = Item(
    name="Paella",
    description="Originating in Valencia, paella is a rice dish prepared\
     with seafood. Of all the foods in Spain, this is the most popular.\
          In this dish, savory yellow rice is combined with tomatoes,\
           onions, peas, shellfish, squid, clams and chicken drumsticks.",
    category=category2
    )
session.add(categoryItem1)
session.commit()

categoryItem2 = Item(
    name="Gazpacho",
    description="Andalusian Gazpacho is a cold soup made of raw blended\
     vegetables. A classic of Spanish cuisine, it originating in the\
      southern region of Andalusia. Gazpacho is widely eaten in Spain\
       and Portugal, particularly during the hot summers, as it is\
        refreshing and cool.",
    category=category2
    )
session.add(categoryItem2)
session.commit()

category3 = Category(name="French Cuisine")
session.add(category3)
session.commit()

categoryItem1 = Item(
    name="Escargot",
    description="Escargots are popular in Spain and Portugal, but are\
     perhaps most known as a part of French cuisine. This escargot\
      recipe calls for canned snails, so there's no need to hunt down\
       any fresh ones, and makes for a sophisticated appetizer at any\
        dinner party.",
    category=category3
    )
session.add(categoryItem1)
session.commit()

categoryItem2 = Item(
    name="Duck Confit",
    description="Duck confit is a French dish made with the whole duck.\
     In Gascony, according to the families perpetuating the tradition of\
      duck confit, all the pieces of duck are used to produce the meal.\
       Each part can have a specific destination in traditional cooking,\
        the neck being used for example in an invigorating soup,\
         the garbure.",
    category=category3
    )
session.add(categoryItem2)
session.commit()

category4 = Category(name="Italian Cuisine")
session.add(category4)
session.commit()

categoryItem1 = Item(
    name="Campania",
    description="Campania extensively produces tomatoes, peppers,\
     spring onions, potatoes, artichokes, fennel, lemons and oranges\
      which all take on the flavor of volcanic soil. The Gulf of Naples\
       offers fish and seafood. Campania is one of the largest producers\
        and consumers of pasta in Italy, especially spaghetti. In the\
         regional cuisine, pasta is prepared in various styles that can\
          feature tomato sauce, cheese, clams and shellfish.",
    category=category4
    )
session.add(categoryItem1)
session.commit()

categoryItem2 = Item(
    name="Risotto",
    description="A northern Italian rice dish cooked in a broth to a\
     creamy consistency. The broth can be derived from meat, fish, or\
      vegetables. Many types of risotto contain butter, wine, and onion.\
       It is one of the most common ways of cooking rice in Italy. Saffron\
        was originally used for flavour and its attractive yellow colour.",
    category=category4
    )
session.add(categoryItem2)
session.commit()

print "DB Updated!"



database_setup.py
import sys
from sqlalchemy import Column, ForeignKey, Integer, String
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import relationship
from sqlalchemy import create_engine

Base = declarative_base()


class User(Base):
    __tablename__ = 'user'
    name = Column(String(250), nullable=False)
    email = Column(String(250), nullable=False)
    picture = Column(String(250), nullable=True)
    id = Column(Integer, primary_key=True)
    status = Column(String(100), default="user")

    @property
    def serialize(self):
        return {
            'name': self.name,
            'id': self.id,
            'email': self.email,
            'picture': self.picture,
            'status': self.status
            }


class Category(Base):
    __tablename__ = 'category'
    name = Column(String(80), nullable=False)
    id = Column(Integer, primary_key=True)
    user_id = Column(Integer, ForeignKey('user.id'))
    user = relationship(User)

    @property
    def serialize(self):
        return {
            'name': self.name,
            'id': self.id,
            'user_id': self.user_id
            }


class Item(Base):
    __tablename__ = 'item'
    name = Column(String(80), nullable=False)
    id = Column(Integer, primary_key=True)
    description = Column(String(500))
    user_id = Column(Integer, ForeignKey('user.id'))
    user = relationship(User)
    category_id = Column(Integer, ForeignKey('category.id'))
    category = relationship(Category)

    @property
    def serialize(self):
        return {
            'name': self.name,
            'id': self.id,
            'description': self.description,
            'category_id': self.category_id,
            'user_id': self.user_id
            }


engine = create_engine('sqlite:////var/www/ItemCatalog/ItemCatalog/main.db')
Base.metadata.create_all(engine)



__init__.py
from flask import Flask, render_template, request, redirect, url_for, jsonify
from flask import flash
from sqlalchemy import create_engine, desc
from sqlalchemy.orm import sessionmaker
from database_setup import Base, User, Category, Item

from flask import session as login_session
import random
import string

from oauth2client.client import flow_from_clientsecrets
from oauth2client.client import FlowExchangeError
import httplib2
import json
from flask import make_response
import requests

app = Flask(__name__)

CLIENT_ID = json.loads(open('/var/www/ItemCatalog/ItemCatalog/client_secrets.json', 'r').read())['web']['client_id']
APPLICATION_NAME = "ItemCatalog"

engine = create_engine('sqlite:////var/www/ItemCatalog/ItemCatalog/main.db?check_same_thread=False')
Base.metadata.bind = engine

DBSession = sessionmaker(bind=engine)
session = DBSession()


# Routing for login
@app.route('/login/')
def login():
    """ Creates a state token, (32) characters long, to prevent request
    forgery."""
    state = ''.join(random.choice(string.ascii_uppercase+string.digits)
                    for x in xrange(32))
    login_session['state'] = state
    return render_template('login.html', STATE=state)


@app.route('/gconnect', methods=['POST'])
def gconnect():
    # Validate state token
    if request.args.get('state') != login_session['state']:
        response = make_response(json.dumps('Invalid state parameter.'), 401)
        response.headers['Content-Type'] = 'application/json'
        return response

    # Obtain authorization code
    code = request.data
    try:
        # Upgrade the authorization code into a credentials object
        oauth_flow = flow_from_clientsecrets('client_secrets.json', scope='')
        oauth_flow.redirect_uri = 'postmessage'
        credentials = oauth_flow.step2_exchange(code)
    except FlowExchangeError:
        response = make_response(json.dumps('Failed to upgrade authorization \
                                            code.'), 401)
        response.headers['Content-Type'] = 'application/json'
        return response

    # Check that the access token is valid
    access_token = credentials.access_token
    url = ('https://www.googleapis.com/oauth2/v1/tokeninfo?access_token=%s'
           % access_token)
    h = httplib2.Http()
    result = json.loads(h.request(url, 'GET')[1])

    # If there was an error in the access token info, abort.
    if result.get('error') is not None:
        response = make_response(json.dumps(result.get('error')), 500)
        response.headers['Content-Type'] = 'application/json'
        return response

    # Verify that the access token is used for the intended user.
    gplus_id = credentials.id_token['sub']
    if result['user_id'] != gplus_id:
        response = make_response(json.dumps("Token's user ID does not match \
                                            given user ID."), 401)
        response.headers['Content-Type'] = 'application/json'
        return response

    # Verify that the access token is valid for this app.
    if result['issued_to'] != CLIENT_ID:
        response = make_response(json.dumps("Token's client ID does not match \
                                            app's."), 401)
        print "Token's client ID does not match app's."
        response.headers['Content-Type'] = 'application/json'
        return response

    stored_access_token = login_session.get('access_token')
    stored_gplus_id = login_session.get('gplus_id')
    if stored_access_token is not None and gplus_id == stored_gplus_id:
        response = make_response(json.dumps('Current user is already \
                                            connected.'), 200)
        response.headers['Content-Type'] = 'application/json'
        return response

    # Store the access token in the session for later use.
    login_session['access_token'] = credentials.access_token
    login_session['gplus_id'] = gplus_id

    # Get user info
    userinfo_url = "https://www.googleapis.com/oauth2/v1/userinfo"
    params = {'access_token': credentials.access_token, 'alt': 'json'}
    answer = requests.get(userinfo_url, params=params)

    data = answer.json()

    login_session['username'] = data['name']
    login_session['picture'] = data['picture']
    login_session['email'] = data['email']

    # See if user exists, if it doesn't create new user
    user_id = getUserID(login_session['email'])
    if not user_id:
        user_id = createUser(login_session)
    login_session['user_id'] = user_id

    output = ''
    output += '<h1>Welcome, '
    output += login_session['username']
    output += '!</h1>'
    output += '<img src="'
    output += login_session['picture']
    output += ' " style = "width: 300px; height: 300px;\
                            border-radius: 150px;-webkit-border-radius: 150px;\
                            -moz-border-radius: 150px;"> '
    flash("You are now logged in as %s" % login_session['username'])
    print "done!"
    return output


# User Helper Functions
def createUser(login_session):
    if(login_session['email'] == "jye0325@gmail.com"):
        newUser = User(name=login_session['username'],
                       email=login_session['email'],
                       picture=login_session['picture'], status='admin')
    else:
        newUser = User(name=login_session['username'],
                       email=login_session['email'],
                       picture=login_session['picture'])
    session.add(newUser)
    session.commit()
    user = session.query(User).filter_by(email=login_session['email']).one()
    return user.id


def getUserInfo(user_id):
    user = session.query(User).filter_by(id=user_id).one()
    return user


def getUserID(email):
    try:
        user = session.query(User).filter_by(email=email).one()
        return user.id
    except:
        return None


# Routing for logout
@app.route('/logout/')
@app.route('/gdisconnect')
def gdisconnect():
    access_token = login_session.get('access_token')
    if access_token is None:
        print 'Access Token is None'
        response = make_response(json.dumps('Current user not connected.'),
                                 401)
        response.headers['Content-Type'] = 'application/json'
        return response
    print 'In gdisconnect access token is %s', access_token
    print 'User name is: '
    print login_session['username']
    url = 'https://accounts.google.com/o/oauth2/revoke?token=%s' \
        % login_session['access_token']
    h = httplib2.Http()
    result = h.request(url, 'GET')[0]
    print 'result is '
    print result
    if result['status'] == '200':
        del login_session['access_token']
        del login_session['gplus_id']
        del login_session['username']
        del login_session['email']
        del login_session['picture']
        response = make_response(json.dumps('Successfully disconnected.'), 200)
        response.headers['Content-Type'] = 'application/json'
        return response
    else:
        response = make_response(json.dumps('Failed to revoke token for given \
                                 user.', 400))
        response.headers['Content-Type'] = 'application/json'
        return response


# Routing for main menu
@app.route('/')
@app.route('/main/')
def mainMenu():
    """Returns a list of categories in the menu pane and the latest 5 added \
    items in the item pane."""
    categories = session.query(Category).all()
    latestItems = session.query(Item).order_by(desc(Item.id)).limit(5)
    if 'username' in login_session:
        return render_template('main.html', categories=categories,
                               latestItems=latestItems, loggedIn=True)
    else:
        return render_template('main.html', categories=categories,
                               latestItems=latestItems)


@app.route('/main/<int:category_id>/')
def categoryItems(category_id):
    """Updates the item pane with a list of items specific to that category"""
    categories = session.query(Category).all()
    items = session.query(Item).filter_by(category_id=category_id)
    if 'username' in login_session:
        return render_template('main.html', categories=categories,
                               items=items, category_id=category_id,
                               loggedIn=True)
    else:
        return render_template('main.html', categories=categories, items=items,
                               category_id=category_id)


@app.route('/main/newCategory/', methods=['GET', 'POST'])
def newCategory():
    """Creates a new category -- Requires Admin Privileges"""
    if 'username' not in login_session:
        return redirect('/login')
    user = session.query(User).filter_by(email=login_session['email']).one()
    if user.status != 'admin':
        return "<script>function myFunction() {alert('Sorry only admin \
                privileges are allowed to add new categories for now.');\
                location.href='http://localhost:5000';}</script><body onload=\
                'myFunction()'>"
    if request.method == 'POST':
        category = Category(name=request.form['name'],
                            user_id=login_session['user_id'])
        session.add(category)
        session.commit()
        flash('Successfully added new category, ' + category.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('newCategory.html')


@app.route('/main/<int:category_id>/edit', methods=['GET', 'POST'])
def editCategory(category_id):
    """Edits a specified category."""
    if 'username' not in login_session:
        return redirect('/login')
    user = session.query(User).filter_by(email=login_session['email']).one()
    category = session.query(Category).filter_by(id=category_id).one()
    if category.user_id == user.id or user.status == 'admin':
        pass
    else:
        return "<script>function myFunction() {alert('Sorry only the creator \
                of this category, or an admin, can edit this.');\
                location.href='http://localhost:5000';}</script>\
                <body onload='myFunction()'>"
    if request.method == 'POST':
        category.name = request.form['name']
        session.add(category)
        session.commit()
        flash('Successfully editted category, ' + category.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('editCategory.html',
                               category=category, id=category_id)


@app.route('/main/<int:category_id>/delete', methods=['GET', 'POST'])
def deleteCategory(category_id):
    """Deletes a specific category."""
    if 'username' not in login_session:
        return redirect('/login')
    user = session.query(User).filter_by(email=login_session['email']).one()
    category = session.query(Category).filter_by(id=category_id).one()
    if category.user_id == user.id or user.status == 'admin':
        pass
    else:
        return "<script>function myFunction() {alert('Sorry only the creator \
                of this category, or an admin, can delete this.');\
                location.href='http://localhost:5000';}</script>\
                <body onload='myFunction()'>"
    if request.method == 'POST':
        session.delete(category)
        session.commit()
        flash('Successfully deleted category, ' + category.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('deleteCategory.html',
                               category=category, id=category_id)


# Routing for item menu
@app.route('/main/<int:category_id>/item/<int:item_id>/')
def itemMenu(category_id, item_id):
    """Returns details about the particular item in the item page"""
    category = session.query(Category).filter_by(category_id=category_id).one()
    item = session.query(Item).filter_by(category_id=category_id,
                                         id=item_id).one()
    return render_template('item.html', category=category, item=item)


@app.route('/main/<int:category_id>/newItem/', methods=['GET', 'POST'])
def newItem(category_id):
    """Creates a new item for a specified category."""
    if 'username' not in login_session:
        return redirect('/login')
    categories = session.query(Category).all()
    if request.method == 'POST':
        item = Item(name=request.form['name'],
                    category_id=request.form['category'],
                    description=request.form['description'],
                    user_id=login_session['user_id'])
        session.add(item)
        session.commit()
        flash('Successfully added new item, ' + item.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('newItem.html', id=category_id,
                               categories=categories)


@app.route('/main/<int:category_id>/item/<int:item_id>/edit/',
           methods=['GET', 'POST'])
def editItem(category_id, item_id):
    """Edit the details about the particular item"""
    if 'username' not in login_session:
        return redirect('/login')
    item = session.query(Item).filter_by(category_id=category_id,
                                         id=item_id).one()
    user = session.query(User).filter_by(email=login_session['email']).one()
    if item.user_id == user.id or user.status == 'admin':
        pass
    else:
        return "<script>function myFunction() {alert('Sorry only the creator \
                of this item, or an admin, can edit this.');\
                location.href='http://localhost:5000';}</script>\
                <body onload='myFunction()'>"
    if request.method == 'POST':
        item.name = request.form['name']
        item.description = request.form['description']
        session.add(item)
        session.commit()
        flash('Successfully editted item, ' + item.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('editItem.html', item=item,
                               id=category_id, item_id=item_id)


@app.route('/main/<int:category_id>/item/<int:item_id>/delete',
           methods=['GET', 'POST'])
def deleteItem(category_id, item_id):
    """Deletes the particular item"""
    if 'username' not in login_session:
        return redirect('/login')
    item = session.query(Item).filter_by(category_id=category_id,
                                         id=item_id).one()
    user = session.query(User).filter_by(email=login_session['email']).one()
    if item.user_id == user.id or user.status == 'admin':
        pass
    else:
        return "<script>function myFunction() {alert('Sorry only the creator \
                of this item, or an admin, can delete this.');\
                location.href='http://localhost:5000';}</script>\
                <body onload='myFunction()'>"
    if request.method == 'POST':
        session.delete(item)
        session.commit()
        flash('Successfully deleted item, ' + item.name)
        return redirect(url_for('mainMenu'))
    else:
        return render_template('deleteItem.html', item=item,
                               id=category_id, item_id=item_id)


@app.route('/status')
def statusChange():
    """For evaluation purposes, this will change the privileges from user to
    admin, or vice-versa."""
    if 'username' not in login_session:
        return redirect('/login')
    user = session.query(User).filter_by(email=login_session['email']).one()
    if user.status == 'admin':
        user.status = 'user'
        session.add(user)
        session.commit()
        flash('Successfully converted, ' + user.name + ' to user privileges')
        return redirect(url_for('mainMenu'))
    elif user.status == 'user':
        user.status = 'admin'
        session.add(user)
        session.commit()
        flash('Successfully converted, ' + user.name + ' to admin privileges')
        return redirect(url_for('mainMenu'))
    else:
        flash('Unknown error. Please try again!')
        return redirect(url_for('mainMenu'))


# JSON Routes
@app.route('/main/JSON')
def mainMenuJSON():
    """Serializes all the categories"""
    category = session.query(Category).all()
    return jsonify(Category=[i.serialize for i in category])


@app.route('/main/<int:category_id>/item/JSON')
def itemMenuJSON(category_id):
    """Serializes all the items within a particular category"""
    category = session.query(Category).all()
    item = session.query(Item).filter_by(category_id=category_id).all()
    return jsonify(Item=[i.serialize for i in item])


@app.route('/main/<int:category_id>/item/<int:item_id>/JSON')
def specificItemJSON(category_id, item_id):
    """Serializes a specific item in a specific category"""
    category = session.query(Category).all()
    item = session.query(Item).filter_by(id=item_id,
                                         category_id=category_id).one()
    return jsonify(Item=item.serialize)


@app.route('/user/JSON')
def userJSON():
    """This is a test JSON to see that users are being created properly"""
    user = session.query(User).all()
    return jsonify(User=[i.serialize for i in user])


if __name__ == '__main__':
    app.secret_key = 'root'
    app.debug = True
    app.run()

